### PR TITLE
TINKERPOP-1889 - Upgrade ws, adding ping and reconnection javascript

### PR DESCRIPTION
**Pull request by spike008t**
https://issues.apache.org/jira/browse/TINKERPOP-1889

- Update ws package
- adding ping interval and pong timeout on websocket to avoid timeout
- auto-reconnection if connection drop

Vote: +1

### CODE REVIEW

**Comment by jorgebay** [comment]
Thanks for your contribution @spike008t !

I didn't have time to check it out this week, I'll look at it next week.

**Comment by spmallette** [comment]
@spike008t thanks for submitting this, i'll have to let @jorgebay speak to the technical aspects of this pull request but administratively speaking:

1. the `tp32` branch is now frozen except for the worst of the worst bug fixes. could you please retarget this to `tp33`?
2. Travis seems to be failing the `gremlin-javascript` build. Do you happen to know if your change is related to that? https://travis-ci.org/apache/tinkerpop/jobs/463928152

**Comment by spike008t** [comment]
@spmallette oh right

1. I'll rebase on tp33. Didn't know that tp32 was frozen.
2. No problem :)

**Comment by jorgebay** [comment]
> Any rejection to add debug package to be able to enable debug by using DEBUG=gremlin:*?

I think we could avoid a new external dependency for the driver. If you are interested in providing debug/info log messages, the `Client` could emit `'log'` events, but feel free to decouple that functionality from this patch.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 1]
```diff
@@ -56,27 +60,16 @@ class Connection {
    * @param {GraphSONWriter} [options.writer] The writer to use.
    * @param {Authenticator} [options.authenticator] The authentication handler to use.
    * @param {Object} [options.headers] An associative array containing the additional header key/values for the initial request.
+   * @param {Boolean} [options.pingEnabled] Setup ping interval
+   * @param {Number} [options.pingInterval] Ping request interval if ping enabled
+   * @param {Number} [options.pongTimeout] Timeout of pong response after sending a ping
+   * @param {Boolean} [options.autoReconnect] Auto reconnect on timeout
+   * @param {Boolean} [options.connectOnStartup] Open websocket on startup
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: We could document the default values for these settings.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 2]
```diff
@@ -96,25 +92,61 @@ class Connection {
     this.isOpen = false;
     this.traversalSource = options.traversalSource || 'g';
     this._authenticator = options.authenticator;
+
+    this._timeoutAutoReconnectionInterval = 500;
+
+    this._pingEnabled = this.options.pingEnabled === false ? false : true;
+    this._pingIntervalDelay = this.options.pingInterval && this.options.pingInterval > 0 ? this.options.pingInterval : pingIntervalDelay;
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: Nit: this condition can be simplified to `this.options.pingInterval || pingIntervalDelay`
(see https://github.com/airbnb/javascript#comparison--unneeded-ternary)

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 3]
```diff
@@ -151,6 +183,64 @@ class Connection {
     });
   }
 
+  _pingHeartbeat() {
+    if (this._pingEnabled === false) {
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: Nit: unnecessary double check.

Comment (spike008t) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js]: removed.

**Comment by spike008t** [comment]
For the reconnection stuff, i did follow what `ws` wiki suggest.
But I'm agree with you by the fact that is not the aim of the ticket.

I'll remove it and create a new ticket for that.

**Comment by jorgebay** [comment]
Landed in d58712c504f41a7953cb004a51f301c6fe99726f.

Thanks @spike008t !!

**Pull request merged**